llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp

```diff
@@ -1787,6 +1787,39 @@ Instruction *InstCombinerImpl::visitAdd(BinaryOperator &I) {
   if (Instruction *Ashr = foldAddToAshr(I))
     return Ashr;
 
+  // Ceiling division by power-of-2:
+  // (X >> log2(N)) + zext(X & (N-1) != 0) --> (X + (N-1)) >> log2(N)
+  // This is valid when adding (N-1) to X doesn't overflow.
+  {
+    Value *X = nullptr, *Cmp = nullptr, *Shift = nullptr;
+    const APInt *ShiftAmt = nullptr, *Mask = nullptr;
+    CmpPredicate Pred;
+
+    // Match: (X >> C) + zext((X & Mask) != 0)
+    // or:    zext((X & Mask) != 0) + (X >> C)
+    if (match(&I, m_c_Add(m_Value(Shift), m_ZExt(m_Value(Cmp)))) &&
+        match(Shift, m_LShr(m_Value(X), m_APInt(ShiftAmt))) &&
+        Shift->hasOneUse() &&
+        match(Cmp, m_ICmp(Pred, m_And(m_Specific(X), m_LowBitMask(Mask)),
+                          m_ZeroInt())) &&
+        Pred == ICmpInst::ICMP_NE &&
+        ShiftAmt && ShiftAmt->uge(1) && ShiftAmt->ult(BitWidth) &&
+        Mask && Mask->popcount() == *ShiftAmt) {
+
+      // Check if X + Mask doesn't overflow
+      Constant *MaskC = ConstantInt::get(X->getType(), *Mask);
+      bool WillNotOverflowUnsigned = willNotOverflowUnsignedAdd(X, MaskC, I);
+
+      if (WillNotOverflowUnsigned) {
+        // (X + Mask) >> ShiftAmt
+        bool WillNotOverflowSigned = willNotOverflowSignedAdd(X, MaskC, I);
+        Value *Add = Builder.CreateAdd(X, MaskC, "", WillNotOverflowUnsigned,
+                                      WillNotOverflowSigned);
```

Comment:
```suggestion
        Value *Add = Builder.CreateNUWAdd(X, MaskC);
```
I don't think it is worthwhile to infer NSW flags here. InstCombine will do it later.